File: javascript/48_Rotate_Image.js
/**
 * @param {number[][]} matrix
 * @return {void} Do not return anything, modify matrix in-place instead.
 */
var rotate = function(matrix) {
    let l = 0, r = matrix.length-1;

    while (l < r) {
        let t = l, b = r;
        for (let i=0; i<(r-l); i++) {
            let topLeft = matrix[t][l + i];
            matrix[t][l + i] = matrix[b - i][l]; 
            matrix[b - i][l] = matrix[b][r - i]; 
            matrix[b][r - i] = matrix[t + i][r]; 
            matrix[t + i][r] = topLeft;
        }
        l ++;
        r --;
    }
};

File: python/48_Rotate_Image.py
'''
Time: O(n^2)
Space: O(1)
'''

class Solution:
    def rotate(self, matrix: List[List[int]]) -> None:
        """
        Do not return anything, modify matrix in-place instead.
        """
        l, r = 0, len(matrix[0])-1  # row pointers  !!! -1

        while l < r:  # layer 
            for i in range(r - l):  # element for each layer
                t, b = l ,r  # column pointers: top, bottom 

                topLeft = matrix[t][l + i]  # save the top left value
                matrix[t][l + i] = matrix[b - i][l]  # bottom left -> top left
                matrix[b - i][l] = matrix[b][r - i]  # bottom right -> bottom left
                matrix[b][r - i] = matrix[t + i][r]  # top right -> bottom right
                matrix[t + i][r] = topLeft  # top left -> top right

            l += 1
            r -= 1